### PR TITLE
Adds support for basic message formatting.

### PR DESCRIPTION
**Pull request by filmil**
Slated for v0.1.1 since this is a backward-compatible API.

Fixes: 56

### CODE REVIEW

**Comment by filmil** [comment]
@kpozin This is work in progress.  I'll need to figure out how to get a nice API from rust, since the varargs formatter in C does not translate well.
Ideas for a type-safe formatter?  Something-something macros?

File: rust_icu_umsg/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+/// The formatting arguments to be used in the call to `UMessageFormat::format`.
+///
+/// According to the [ICU documentation for `umsg_format`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/umsg_8h.html#a90a4b5fe778754e5da52f7c2e5fd6048):
+///
+/// > for all numeric arguments double is assumed unless the type is explicitly
+/// > integer (long).  All choice format arguments must be of type double.
+///
+/// Additionally, we determined by inspection that the string format is necessarily
+/// `rust_icu_ustring::UChar`.
+pub enum FmtArg {
```
Comment (kpozin) [inline rust_icu_umsg/src/lib.rs]: Please implement `From<f64>`, `From<i32>`, etc., for `FmtArg`. For the string variants, please implement conversions from `&str` and `String`.

File: rust_icu_umsg/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        common::Error::ok_or_warning(status)?;
+        parse_ok(parse_status)?;
+        Ok(UMessageFormat { rep })
+    }
+
+    /// Formats a new message based on this formatter.
+    ///
+    /// Implements `umsg_format`.
+    /// Implements `umsg_vformat`.
+    pub fn format(&self, args: &[FmtArg]) -> Result<String, common::Error> {
```
Comment (kpozin) [inline rust_icu_umsg/src/lib.rs]: `args` should be `&[impl Into<FmtArg>]` or something similar.

File: rust_icu_umsg/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+) -> Result<String, common::Error> {
+    const CAP: usize = 1024;
+    let mut status = common::Error::OK_CODE;
+    let mut result = ustring::UChar::new_with_capacity(CAP);
+
+    let total_size = {
+        args.with_copy(|va_list| {
+            // Overlay a __va_list_tag on top of our va_list.
+            //
+            // This is terribly unsafe, and hinges on the assumption that the memory layout of
```
Comment (kpozin) [inline rust_icu_umsg/src/lib.rs]: Nit: column width

File: rust_icu_umsg/src/lib.rs (hunk 1 above)
Comment (filmil) [reply rust_icu_umsg/src/lib.rs]: I rethought the formatting, and looks like `FmtArg` is not needed.

File: rust_icu_umsg/src/lib.rs (hunk 2 above)
Comment (filmil) [reply rust_icu_umsg/src/lib.rs]: See above: not needed.

File: rust_icu_umsg/src/lib.rs (hunk 3 above)
Comment (filmil) [reply rust_icu_umsg/src/lib.rs]: Hm, 100 cols isn't so bad.  I'd keep?

File: rust_icu_umsg/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+///   )?;
+///
+///   let fmt = umsg::UMessageFormat::try_from(&msg, &loc)?;
+///   let hello = ustring::UChar::try_from("Hello! Добар дан!")?;
+///   let _ = message_format!(&fmt);
+///   let result = umsg::message_format!(
+///     fmt,
+///     { 43.4 => Double },
+///     { 31337 => Integer},
+///     { hello => String},
```
Comment (kpozin) [inline rust_icu_umsg/src/lib.rs]: As a future improvement, we might want to support `&str` literals here.

Comment (filmil) [reply rust_icu_umsg/src/lib.rs]: I am still not sure about this API... I guess we will see.

**Pull request merged**